kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraWindowedTable.java

```diff
@@ -89,24 +110,59 @@ public static CassandraWindowedTable create(
     // Cassandra does not support filtering on a composite key column if
     // the previous columns in the composite are not equality filters
     // in the table below, for example, we cannot filter on WINDOW_START
-    // unless both PARTITION_KEY and DATA_KEY are equality filters because
+    // unless DATA_KEY is an equality filter, or vice versa -- because
     // of the way SSTables are used in Cassandra this would be inefficient
     //
-    // until we figure out a better data model (perhaps segmenting the
-    // stores using an additional column to reduce the scan overhead -
-    // this would also help us avoid having a partitioning key that is
-    // too small in cardinality) we just filter the results to match the
-    // time bounds
+    // Until we figure out a better data model we just filter on the
+    // DATA_KEY and then post-filter the results to match the time bounds.
+    // Although we may fetch results that don't strictly fall within the
+    // query bounds, the extra data is limited to at most twice the
+    // segment interval, since the segments already narrow down the time
+    // range although at a more coarse-grained size
+    //
+    // This is probably sufficient for now, as the key range fetches --
+    // especially the bounded key-range fetch, ie fetchRange -- are both
+    // relatively quite uncommon in Streams applications. Note that the
+    // DSL only uses point or single-key lookups, and even among PAPI
+    // users, key-range queries are typically rare due to several factors
+    // (mainly the unpredictable ordering, as well as unidentifiable
+    // bounds for the fetchRange queries, etc)
     LOG.info("Creating windowed data table {} in remote store.", name);
     final CreateTableWithOptions createTable = spec.applyOptions(createTable(name));
 
     client.execute(createTable.build());
     client.awaitTable(name).await(Duration.ofSeconds(60));
 
+    // TODO: consolidate the #init method into the constructor and create the
+    //  LwtWriterFactory before preparing the statements so we can just plug
+    //  in the epoch directly without having to bind it later/on each request
+    final var initSegment = client.prepare(
+        QueryBuilder.insertInto(name)
+            .value(PARTITION_KEY.column(), bindMarker(PARTITION_KEY.bind()))
+            .value(SEGMENT_ID.column(), bindMarker(SEGMENT_ID.bind()))
+            .value(ROW_TYPE.column(), METADATA_ROW.literal())
+            .value(DATA_KEY.column(), DATA_KEY.literal(ColumnName.METADATA_KEY))
+            .value(WINDOW_START.column(), WINDOW_START.literal(METADATA_TS))
+            .value(EPOCH.column(), bindMarker(EPOCH.bind()))
+            .ifNotExists()
+            .build()
+    );
+
+    // Note: it is safe to ignore the epoch here as we should only ever drop a segment
```

Comment:
I think we have risk of data revival though. Here's the situation:
1. writer A attempts to write to segment 1 kv pair (X,Y)
2. writer A takes a walk (GC, IDK)
3. writer B is assigned as active
4. writer B successfully writes to segment 1 kv pair (X,Y)
5. time passes
6. writer B successfully expires segment 1
7. writer A wakes up and successfully writes kv pair (X, Y) to segment 1, reviving that data

We need to make sure that 7 fails, and if we check the epoch per-segment then it isn't possible. One suggestion is to have expire segment first delete everything, and then atomically write a tombstone metadata row. Not sure how to deal with expiring these tombstones though...

Perhaps we can also just be "OK" with this race condition because that expired segment will never be queried.

Reply:
Added a TODO and filed a ticket for this. Nice catch